library/std/src/sys/unix/fs.rs

```diff
@@ -1648,55 +1667,155 @@ mod remove_dir_impl {
+    }
+
+    fn readdir_open_path(path: &CStr) -> io::Result<(DirComponent, CachedReadDir)> {
+        let dir_fd = openat_nofollow_dironly(None, path)?;
+
+        // use fstat() to get the inode of the directory
+        let mut stat = unsafe { mem::zeroed() };
+        cvt(unsafe { fstat64(dir_fd.as_raw_fd(), &mut stat) })?;
+        let (readdir, raw_fd) = fdreaddir(dir_fd)?;
+        Ok((
+            DirComponent { name: CString::new("")?, ino: stat.st_ino },
+            CachedReadDir { readdir, raw_fd },
+        ))
+    }
+
+    fn readdir_open_child(
+        readdir: &CachedReadDir,
+        child: &DirEntry,
+    ) -> io::Result<(DirComponent, CachedReadDir)> {
+        let dir_fd = openat_nofollow_dironly(Some(readdir.raw_fd), child.name_cstr())?;
+        let (readdir, raw_fd) = fdreaddir(dir_fd)?;
+        Ok((
+            DirComponent { name: child.name_cstr().into(), ino: child.entry.d_ino },
+            CachedReadDir { readdir, raw_fd },
+        ))
+    }
+
+    fn readdir_reopen_parent(
+        dir: &CachedReadDir,
+        expected_parent_dir: &DirComponent,
+    ) -> io::Result<CachedReadDir> {
+        let parent_dir_fd = openat_nofollow_dironly(Some(dir.raw_fd), unsafe {
+            CStr::from_bytes_with_nul_unchecked(b"..\0")
         })?;
+        let mut stat = unsafe { mem::zeroed() };
+        cvt(unsafe { fstat64(parent_dir_fd.as_raw_fd(), &mut stat) })?;
+        // Make sure that the reopened parent directory has the same inode as when we visited it descending
+        // the directory tree. More detailed risk analysis TBD.
+        if expected_parent_dir.ino != stat.st_ino {
+            return Err(io::Error::new(
+                io::ErrorKind::Uncategorized,
+                "parent directory inode does not match",
+            ));
+        }
+        let (readdir, raw_fd) = fdreaddir(parent_dir_fd)?;
+        Ok(CachedReadDir { readdir, raw_fd })
+    }
+
+    fn remove_dir_all_loop(root: &Path) -> io::Result<()> {
+        // all ancestor names and inodes from the deletion root directory to the parent of the currently processed
+        // directory
+        let mut parent_dir_components = Vec::new();
+        // cache of up to MAX_OPEN_FDS ancestor ReadDirs and associated file descriptors
+        let mut readdir_cache = VecDeque::with_capacity(MAX_OPEN_FDS);
+        // the directory name, inode pair and ReadDir currently being processed
+        let (mut current_dir_component, mut current_readdir) = readdir_open_path(&cstr(root)?)?;
+        loop {
+            while let Some(child) = current_readdir.readdir.next() {
+                let child = child?;
+                if !unlink_direntry(&child, current_readdir.raw_fd)? {
+                    // Descend into this child directory
+
+                    let (child_dir_compoment, child_readdir) =
+                        readdir_open_child(&current_readdir, &child)?;
+                    parent_dir_components.push(current_dir_component);
```

Comment:
We could do `try_reserve` here and return `ENOMEM` when it fails.

Reply:
We could still fail on CString allocation for the directory name as well though.

Reply:
The CString allocation for the directory name could fail as well, haven't check if that has something like `try_reserve()`.

Reply:
Indirectly via `from_vec_unchecked`